src/printer/tvmscript_unified_printer.h

```diff
@@ -0,0 +1,656 @@
+  ExprDoc IndexWith(IndexType&&... args);
+
+  ExprDoc() : ExprDoc(make_object<ExprDocNode>()){};
+  TVM_DEFINE_MUTABLE_NOTNULLABLE_OBJECT_REF_METHODS(ExprDoc, Doc, ExprDocNode);
+};
+
+// Literal Value Doc
+
+class LiteralValueDocNode : public ExprDocNode {
+ public:
+  ObjectRef value;  // Can only be FloatImm/IntImm/StringImm/String
+
+  static constexpr const char* _type_key = "script.Docs.LiteralValueDoc";
+  TVM_DECLARE_FINAL_OBJECT_INFO(LiteralValueDocNode, ExprDocNode);
+
+ private:
+  ObjectRef value_;
+};
+
+class LiteralValueDoc : public ExprDoc {
+ public:
+  LiteralValueDoc() : LiteralValueDoc(make_object<LiteralValueDocNode>()){};
+  LiteralValueDoc(IntImm val) : LiteralValueDoc(static_cast<ObjectRef>(val)) {}
+  LiteralValueDoc(FloatImm val) : LiteralValueDoc(static_cast<ObjectRef>(val)) {}
+  LiteralValueDoc(tir::StringImm val) : LiteralValueDoc(static_cast<ObjectRef>(val)) {}
+  LiteralValueDoc(String val) : LiteralValueDoc(static_cast<ObjectRef>(val)) {}
+
+  TVM_DEFINE_MUTABLE_NOTNULLABLE_OBJECT_REF_METHODS(LiteralValueDoc, ExprDoc, LiteralValueDocNode);
+
+ protected:
+  explicit LiteralValueDoc(ObjectRef value) {
+    auto node = make_object<LiteralValueDocNode>();
+    node->value = std::move(value);
+    data_ = std::move(node);
+  }
+};
+
+// Special Constant Doc
+
+class ConstDocNode : public ExprDocNode {
+ public:
+  enum class ConstKind { TIRBuilder, RelaxBuilder, None };
+
+  ConstKind kind;
+
+  ConstDocNode(ConstKind kind) : kind(kind) {}
+
+  static constexpr const char* _type_key = "script.Docs.ConstDoc";
+  TVM_DECLARE_FINAL_OBJECT_INFO(ConstDocNode, ExprDocNode);
+};
+
+class ConstDoc : public ExprDoc {
+ public:
+  TVM_DEFINE_MUTABLE_NOTNULLABLE_OBJECT_REF_METHODS(ConstDoc, ExprDoc, ConstDocNode);
+};
+
+// Identifier Doc
+
+class IdentifierDocNode : public ExprDocNode {
+ public:
+  String name;
+
+  static constexpr const char* _type_key = "script.Docs.IdentidierDoc";
+  TVM_DECLARE_FINAL_OBJECT_INFO(IdentifierDocNode, ExprDocNode);
+};
+
+class IdentifierDoc : public ExprDoc {
+ public:
+  IdentifierDoc(String name) {
+    auto node = make_object<IdentifierDocNode>();
+    node->name = std::move(name);
+    data_ = std::move(node);
+  };
+
+  IdentifierDoc() : IdentifierDoc(make_object<IdentifierDocNode>()){};
+  TVM_DEFINE_MUTABLE_NOTNULLABLE_OBJECT_REF_METHODS(IdentifierDoc, ExprDoc, IdentifierDocNode);
+};
+
+// Attr Access Doc
+
+class AttrAccessDocNode : public ExprDocNode {
+ public:
+  ExprDoc value;
+  IdentifierDoc attr;
+
+  static constexpr const char* _type_key = "script.Docs.AttrAccessDoc";
+  TVM_DECLARE_FINAL_OBJECT_INFO(AttrAccessDocNode, ExprDocNode);
+};
+
+class AttrAccessDoc : public ExprDoc {
+ public:
+  AttrAccessDoc() : AttrAccessDoc(make_object<AttrAccessDocNode>()){};
+  TVM_DEFINE_MUTABLE_NOTNULLABLE_OBJECT_REF_METHODS(AttrAccessDoc, ExprDoc, AttrAccessDocNode);
+};
+
+// Index Doc
+
+class IndexDocNode : public ExprDocNode {
+ public:
+  ExprDoc value;
+  Array<ExprDoc> indices;
```

Comment:
```suggestion
  Array<RangeDoc> indices;
```

where RangeDoc could be:

```
Optional<ExprDoc> start;
Optional<ExprDoc> end;
Optional<ExprDoc> step;
```